src/vs/workbench/contrib/chat/common/tools/projectTreeTool.ts

```diff
@@ -0,0 +1,139 @@
+ *  Copyright (C) 2025 Posit Software, PBC. All rights reserved.
+ *  Licensed under the Elastic License 2.0. See LICENSE.txt for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { CancellationToken } from '../../../../../base/common/cancellation.js';
+import { localize } from '../../../../../nls.js';
+import { IWorkspaceContextService } from '../../../../../platform/workspace/common/workspace.js';
+import { IExplorerService } from '../../../files/browser/files.js';
+import { CountTokensCallback, IToolData, IToolImpl, IToolInvocation, IToolResult } from '../languageModelToolsService.js';
+import { ExplorerItem } from '../../../files/common/explorerModel.js';
+import { SortOrder } from '../../../files/common/files.js';
+import { IToolInputProcessor } from './tools.js';
+
+/**
+ * Represents either a file (string) or a directory (tuple with string and children).
+ */
+export type DirectoryItem = string | [string, DirectoryItem[]];
+
+export const ExtensionProjectTreeId = 'positron_getProjectTree';
+export const InternalProjectTreeId = `${ExtensionProjectTreeId}_internal`;
+export const ProjectTreeToolData: IToolData = {
+	id: InternalProjectTreeId,
+	displayName: localize('chat.tools.getProjectTree', "Get Project Tree"),
+	modelDescription: `Construct a tree of the directories and files in the project.`,
+	tags: ['positron-assistant']
+};
+
+export class ProjectTreeTool implements IToolImpl {
+	constructor(
+		@IWorkspaceContextService private readonly _workspaceContextService: IWorkspaceContextService,
+		@IExplorerService private readonly _explorerService: IExplorerService,
+	) { }
+
+	async invoke(invocation: IToolInvocation, countTokens: CountTokensCallback, token: CancellationToken): Promise<IToolResult> {
+		const workspaceFolders = this._workspaceContextService.getWorkspace().folders;
+		if (workspaceFolders.length === 0) {
+			return {
+				content: [],
+				toolResultMessage: 'No workspace folders found.'
+			};
+		}
+
+		const workspaceTrees: DirectoryItem[][] = [];
+		const treeErrors: string[] = [];
+		const sortOrder = this._explorerService.sortOrderConfiguration.sortOrder;
+
+		for (const workspaceFolder of workspaceFolders) {
+			const explorerRoot = this._explorerService.findClosestRoot(workspaceFolder.uri);
+			if (!explorerRoot) {
+				treeErrors.push(`No explorer root found for ${workspaceFolder.name}`);
+				continue;
+			}
+			try {
+				const tree = await constructDirectoryTree(explorerRoot, sortOrder);
+				if (!tree) {
+					continue;
+				}
+				workspaceTrees.push(tree);
+			} catch (error) {
+				treeErrors.push(`Failed to generate tree for ${workspaceFolder.name}: ${error}`);
+			}
+		}
+
+		return {
+			content: workspaceTrees.map(dirTree => ({ kind: 'text', value: JSON.stringify(dirTree) })),
+			toolResultMessage: treeErrors.length > 0 ? treeErrors.join('\n') : undefined
```

Comment:
How are you thinking about when to include text in `content` versus `toolResultMessage` versus throwing an error?

Reply:
I've now modified the error handling to throw, as it's pretty unexpected if we can't generate the project tree for some reason. Instead of returning errors in the toolResultMessage here, we throw just prior if any errors occur.